Google Photos will not allow preview of editing in Safari.

mattwoodrow opened the pull request:
#### cbd73ab873ddc90f1857c70d2214dedd586ae661
<pre>
Google Photos will not allow preview of editing in Safari.
<a href="https://bugs.webkit.org/show_bug.cgi?id=259263">https://bugs.webkit.org/show_bug.cgi?id=259263</a>
&lt;rdar://107687640&gt;

Reviewed by Dean Jackson.

OffscreenCanvas can asynchronously push frames to the layer associated with the HTMLCanvasElement that is being controlled.
This happens using a GraphicsLayerAsyncContentsDisplayDelegate, created by the GraphicsLayer for the HTMLCanvasElement.

If style changes that result in the recreation of the GraphicsLayer, the OffscreenCanvas continues to push frames to the original
delegate, and they get dropped.

This makes ImageBufferPipe always ask the current GraphicsLayer to make sure we have a valid delegate, providing the delegate
that was used last time. If the GraphicsLayer has changed, it can reconfigure the delegate to make sure it&apos;s targetting the right layer.

We need to reconfigure the delegate, rather than recreating one, since the delegate owns the &apos;current&apos; frame and we need to ensure this
is retained. Otherwise compositing wouldn&apos;t have a frame until the next time the OffscreenCanvas draws (which could be never).

* LayoutTests/fast/canvas/offscreen-toggle-display-expected.html: Added.
* LayoutTests/fast/canvas/offscreen-toggle-display.html: Added.
* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(WebCore::GraphicsLayer::createAsyncContentsDisplayDelegate):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
* Source/WebCore/platform/graphics/GraphicsLayerContentsDisplayDelegate.h:
* Source/WebCore/platform/graphics/ImageBufferPipe.cpp:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::contentsLayerIDForModel const):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/cocoa/GraphicsLayerAsyncContentsDisplayDelegateCocoa.h:
* Source/WebCore/platform/graphics/ca/cocoa/GraphicsLayerAsyncContentsDisplayDelegateCocoa.mm:
(WebCore::GraphicsLayerAsyncContentsDisplayDelegateCocoa::updateGraphicsLayerCA):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.mm:
(WebKit::GraphicsLayerCARemote::createAsyncContentsDisplayDelegate):

Canonical link: <a href="https://commits.webkit.org/266120@main">https://commits.webkit.org/266120@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/af779c333cd4ae781c8a9a49e10341515ce08922

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12926 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13250 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13580 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14664 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12337 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12994 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15754 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13272 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15033 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13092 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13802 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10939 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15115 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11090 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11692 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/18759 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12165 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11859 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15067 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12323 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10218 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11585 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/11559 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3170 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15903 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12164 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

grorg reviewed (approved):
r+ but the new test seems to fail on macOS.

mattwoodrow committed:
Google Photos will not allow preview of editing in Safari.
https://bugs.webkit.org/show_bug.cgi?id=259263
<rdar://107687640>

Reviewed by Dean Jackson.

OffscreenCanvas can asynchronously push frames to the layer associated with the HTMLCanvasElement that is being controlled.
This happens using a GraphicsLayerAsyncContentsDisplayDelegate, created by the GraphicsLayer for the HTMLCanvasElement.

If style changes that result in the recreation of the GraphicsLayer, the OffscreenCanvas continues to push frames to the original
delegate, and they get dropped.

This makes ImageBufferPipe always ask the current GraphicsLayer to make sure we have a valid delegate, providing the delegate
that was used last time. If the GraphicsLayer has changed, it can reconfigure the delegate to make sure it's targetting the right layer.

We need to reconfigure the delegate, rather than recreating one, since the delegate owns the 'current' frame and we need to ensure this
is retained. Otherwise compositing wouldn't have a frame until the next time the OffscreenCanvas draws (which could be never).

* LayoutTests/fast/canvas/offscreen-toggle-display-expected.html: Added.
* LayoutTests/fast/canvas/offscreen-toggle-display.html: Added.
* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(WebCore::GraphicsLayer::createAsyncContentsDisplayDelegate):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
* Source/WebCore/platform/graphics/GraphicsLayerContentsDisplayDelegate.h:
* Source/WebCore/platform/graphics/ImageBufferPipe.cpp:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::contentsLayerIDForModel const):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/cocoa/GraphicsLayerAsyncContentsDisplayDelegateCocoa.h:
* Source/WebCore/platform/graphics/ca/cocoa/GraphicsLayerAsyncContentsDisplayDelegateCocoa.mm:
(WebCore::GraphicsLayerAsyncContentsDisplayDelegateCocoa::updateGraphicsLayerCA):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.mm:
(WebKit::GraphicsLayerCARemote::createAsyncContentsDisplayDelegate):

Canonical link: https://commits.webkit.org/266120@main
